server/src/main/java/org/elasticsearch/cluster/metadata/MappingMetadata.java

```diff
@@ -41,36 +45,53 @@
 
     private final boolean routingRequired;
 
+    private final String digest;
+
     public MappingMetadata(DocumentMapper docMapper) {
         this.type = docMapper.type();
         this.source = docMapper.mappingSource();
         this.routingRequired = docMapper.routingFieldMapper().required();
+        this.digest = computeDigest(source);
     }
 
-    @SuppressWarnings("unchecked")
     public MappingMetadata(CompressedXContent mapping) {
         this.source = mapping;
         Map<String, Object> mappingMap = XContentHelper.convertToMap(mapping.compressedReference(), true).v2();
         if (mappingMap.size() != 1) {
             throw new IllegalStateException("Can't derive type from mapping, no root type: " + mapping.string());
         }
         this.type = mappingMap.keySet().iterator().next();
-        this.routingRequired = routingRequired((Map<String, Object>) mappingMap.get(this.type));
+        this.routingRequired = routingRequired((Map<?, ?>) mappingMap.get(this.type));
+        this.digest = computeDigest(mapping);
     }
 
-    @SuppressWarnings("unchecked")
     public MappingMetadata(String type, Map<String, Object> mapping) {
         this.type = type;
+        String mappingAsString = Strings.toString((builder, params) -> builder.mapContents(mapping));
         try {
-            this.source = new CompressedXContent((builder, params) -> builder.mapContents(mapping));
+            this.source = new CompressedXContent(mappingAsString);
```

Comment:
Can't we use `this.source = new CompressedXContent((builder, params) -> builder.mapContents(mapping));` here still and compute the digest as we compress for example since we stream the bytes in that case anyway?

Reply:
On that note, I wonder if we should just make the `Sha256` value a part of the `CompressedXContent`? I see how we do some magic here to make sure we build a deterministic version based on a map, but we could have that by just adding a static constructor to this thing? For BwC we can compute the CRC or SHA1 dynamically where/if needed I guess.
Also, that would allow us to make the `equals` in `CompressedXContent` almost free by just turning it into a string comparison :) WDYT?

Reply:
I like this idea and in fact I did think about this while working on this. I will try this out and see whether it is not too difficult.

Reply:
I did the first part of this: c7705231c9285e75c809619bb80bc0176c131657
I will work on making the sha256 really part of of `CompressedXContent`, just like crc32 checksum.
Just to double check, do we want to keep the `crc32` checksum? I think so.

Reply:
> Just to double check, do we want to keep the crc32 checksum? I think so.

Not sure tbh. It seems we're only using it to speed up comparisons and as a `hashCode`. Both of which we can have from the sha256 value can't we?

Reply:
Right, I thought there was another usage of the crc32, but it only used for hashcode and checking consistency in assertions. Both can be done with sha256 too. I will replace crc32 completely with sha256.

Reply:
I pushed: 1961453aa16caa9468f541b975a047a15ae56e01